include/ensmallen_bits/callbacks/progress_bar.hpp

```diff
@@ -164,8 +164,43 @@ class ProgressBar
     output << "] " << progress << "% - ETA: " << (size_t) stepTimer.toc() *
         (epochSize - step + 1) % 60 << "s - loss: " <<
         objective / (double) step <<  "\r";
-    output.flush();
 
+    if (optimizer.MaxIterations() < function.NumFunctions() &&
```

Comment:
I don't think that this is the right way to check this condition.  This will fail if the termination happens during the second epoch.  Here's some example code:

```
#include <ensmallen.hpp>

int main()
{
  arma::mat data(10, 1000, arma::fill::randu);
  arma::Row<size_t> responses(1000);
  for (size_t i = 0; i < 500; ++i)
    responses[i] = 0;
  for (size_t i = 500; i < 1000; ++i)
    responses[i] = 1;

  ens::test::LogisticRegressionFunction<> lrf(data, responses, 0.1);

  ens::Adam adam;
  adam.BatchSize() = 1;
  adam.MaxIterations() = 1500;
  arma::mat coordinates = arma::randu<arma::mat>(1, 11);
  adam.Optimize(lrf, coordinates, ens::ProgressBar());
}
```

If you try running that on this branch, you'll see:

```
$ ./test && echo ""
Epoch 1/2
1000/1000 [====================================================================================================] 100% - 0s 0ms/step - loss: 1.23509
Epoch 2/2
500/1000 [==================================================>.................................................] 50% - ETA: 0s - loss: 0.836281
$
```

(The `&& echo ""` is necessary because the last `\r` is printed unnecessarily.)

I wonder if we need to do this in a separate way, where a final call to this callback happens when the method terminates.  Perhaps in the destructor of `ProgressBar`?  I'm not sure if that could work, but it might.